fms/modules/speculator.py

```diff
@@ -0,0 +1,89 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from fms.modules.layernorm import LayerNormParameterized
+
+
+class Speculator(nn.Module):
+    def __init__(self, emb_dim=4096, vocab_size=32000, n_heads=4):
+        super().__init__()
+        self.nheads = n_heads
+        self.emb_dim = emb_dim
+        self.vsize = vocab_size
+        self.emb = nn.ModuleList([nn.Embedding(vocab_size, emb_dim) for _ in range(n_heads)])
+        self.proj = nn.ModuleList([nn.Linear(emb_dim * 2, emb_dim, bias=False) for _ in range(n_heads)])
+        self.head = nn.ModuleList([nn.Linear(emb_dim, vocab_size, bias=False) for _ in range(n_heads)])
+        self.ln = nn.ModuleList(
+            [LayerNormParameterized(emb_dim, elementwise_shift=True, elementwise_scale=True) for _ in range(n_heads)]
+        )
+        self.a = nn.GELU()
+        self.reset_params()
+
+    def reset_params(self):
+        for m in self.modules():
+            if isinstance(m, nn.Embedding) or isinstance(m, nn.Linear):
+                nn.init.trunc_normal_(m.weight, 0, 1 / self.emb_dim**0.5)
+            elif isinstance(m, LayerNormParameterized):
+                m.weight.data.fill_(1)
+                m.bias.data.zero_()
+                
+    def generate_tree(self, state, ind, topk=[5,4,3], k=5):
+        '''
+        FOR INFERENCE
+        ----
+        Generate tree of candidate sequences given latest base model embedding (state) and chosen token (ind).
+        Topk indicates # of tree "branches" at each head. 
+        k pares down the candidate list from prod(topk) to the top k most confident. 
```

Comment:
maybe rename the second k to another letter, so as to not have the top-k for each head, and then the k best for the total combination?